### PR TITLE
Feat(validation): Implement ValidatableValue for yaml_parser values

### PR DESCRIPTION
**Pull request by ClausHolbechArista**
- Implements the ValidatableValue trait for yaml_parser::Value and associated types.
- Adds validate_yaml() to the StoreValidateInput trait.

This is not exposed to Python so it will not be used for pyavd-utils library (yet).

### CODE REVIEW

File: rust/validation/src/validatable/yaml_parser_impl.rs [hunk 1]
```diff
@@ -0,0 +1,818 @@
...
+                Integer::BigIntStr(s) => Some(s.clone()),
+            },
+            Value::Float(f) => Some(Cow::Owned(f.to_string())),
+            // Using Title case to match Python behavior
+            Value::Bool(b) => Some(Cow::Borrowed(if *b { "True" } else { "False" })),
+            _ => None,
+        }
+    }
+
+    fn as_i64(&self) -> Option<i64> {
```
Comment (gmuloc) [inline rust/validation/src/validatable/yaml_parser_impl.rs:62]: as_i64 and is_int can give contradicting information for big integers (e.g. I128, U128 and BigIntStr)

`9223372036854775808` -> Yaml Parser says it is an int but as_i64 says it is None. This may be acceptable but a weird corner case leading to 

```
Expected type int but got int
```

File: rust/validation/src/validatable/yaml_parser_impl.rs [hunk 2]
```diff
@@ -0,0 +1,818 @@
...
+            Value::String(_) => Type::Str,
+            Value::Sequence(_) => Type::List,
+            Value::Mapping(_) => Type::Dict,
+        }
+    }
+
+    // === Coercion builders ===
+    // These preserve the span from the original node.
+
+    fn coerce_null(&self) -> Self::Coerced {
```
Comment (gmuloc) [inline rust/validation/src/validatable/yaml_parser_impl.rs:127]: here we are dropping comments, anchor, tags,... from the original node. I think it is ok as we are not using anything else today but that could be something to revisit.

File: rust/validation/src/validatable/yaml_parser_impl.rs [hunk 3]
```diff
@@ -0,0 +1,818 @@
...
+                let pair_span = yaml_parser::Span::new(key_span.start..value.span.end);
+                let key_node = Node::new(Value::String(Cow::Owned(key)), key_span);
+                MappingPair::new(pair_span, key_node, value)
+            })
+            .collect();
+        Node::new(Value::Mapping(pairs), self.span)
+    }
+
+    fn clone_to_coerced(&self) -> Self::Coerced {
+        self.clone().into_owned()
```
Comment (gmuloc) [inline rust/validation/src/validatable/yaml_parser_impl.rs:160]: this would keep the Node metadata so we could end up with mixed flavors of Nodes. (but as I said, this seems ok for now)

File: rust/validation/src/validatable/yaml_parser_impl.rs [hunk 4]
```diff
@@ -0,0 +1,818 @@
...
+            self.span,
+        )
+    }
+
+    fn coerce_mapping(&self, items: Vec<(String, Self::Coerced)>) -> Self::Coerced {
+        // Convert string keys to Node keys
+        let pairs: Vec<MappingPair<'static>> = items
+            .into_iter()
+            .map(|(key, value)| {
+                let key_span = yaml_parser::Span::at(value.span.start);
```
Comment (gmuloc) [inline rust/validation/src/validatable/yaml_parser_impl.rs]: feels like we are losing information here

File: rust/validation/src/validatable/yaml_parser_impl.rs [hunk 5]
```diff
@@ -0,0 +1,818 @@
...
+                let pair_span = yaml_parser::Span::new(key_span.start..value.span.end);
```
Comment (gmuloc) [inline rust/validation/src/validatable/yaml_parser_impl.rs]: this feels weird to use `key_span.start` which is actually `value.span.start` here as it gives some wrong impression but I guess thats what we are passing today and it looks ok from usage

File: rust/validation/src/validatable/yaml_parser_impl.rs [hunk 6]
```diff
@@ -0,0 +1,818 @@
...
+            Value::Sequence(seq) => FV::List(
+                seq.iter()
+                    .map(|item| item.node.to_feedback_value())
+                    .collect(),
+            ),
+            Value::Mapping(map) => FV::Dict(
+                map.iter()
+                    .filter_map(|pair| {
+                        pair.key
+                            .as_str()
```
Comment (gmuloc) [inline rust/validation/src/validatable/yaml_parser_impl.rs]: this could cause some weirdness for bool key as as_str returns "True"/"False" when coerce_key_to_string wuld return "true"/"false"

Comment (gmuloc) [reply rust/validation/src/validatable/yaml_parser_impl.rs]: but we don't use bool keys so we should survive

Comment (gmuloc) [reply rust/validation/src/validatable/yaml_parser_impl.rs]: hmm given the comment below I wonder if that would not be better though to use the same function

Comment (gmuloc) [reply rust/validation/src/validatable/yaml_parser_impl.rs]: showing the inconsistency
```rust
  #[test]
  fn yaml_edge_case_feedback_value_uses_title_case_bool_mapping_key() {
      let mapping = Node::new(
          Value::Mapping(vec![MappingPair::new(
              make_span(),
              Node::new(Value::Bool(true), make_span()),
              string_node("enabled"),
          )]),
          make_span(),
      );

      let feedback_value = mapping.to_feedback_value();

      assert_eq!(
          feedback_value,
          crate::feedback::Value::Dict(std::collections::HashMap::from([(
              "True".to_owned(),
              crate::feedback::Value::Str("enabled".to_owned()),
          )]))
      );

      assert_eq!(
          mapping
              .get("true")
              .and_then(|node| node.as_str())
              .as_deref(),
          Some("enabled")
      );

      assert!(mapping.get("True").is_none());
  }
```

File: rust/validation/src/validatable/yaml_parser_impl.rs (hunk 2 above)
Comment (ClausHolbechArista) [reply rust/validation/src/validatable/yaml_parser_impl.rs:127]: Correct, the properties and trailing comment are relatively expensive to clone so unless we need them I think we should avoid that.
I hope we can end up later at a place where we don't have to rebuild all the data for coercion.

Comment (ClausHolbechArista) [reply rust/validation/src/validatable/yaml_parser_impl.rs:127]: added a comment

File: rust/validation/src/validation/store.rs [hunk 7]
```diff
@@ -31,6 +33,13 @@ pub struct InputValidationResult<T> {
     pub document: ValidationOutput<T>,
 }
 
+#[derive(Debug)]
+/// Result of validating a YAML input source that may contain multiple documents.
+pub struct YamlValidationResult<T> {
```
Comment (gmuloc) [inline rust/validation/src/validation/store.rs:38]: would it make sense to have only one InputValidationResult and change the document to documents?

File: rust/validation/src/validation/store.rs [hunk 8]
```diff
@@ -120,9 +171,124 @@ pub enum StoreValidateError {
...
     use super::*;
+    use crate::feedback::Feedback;
     use crate::feedback::ParseDiagnosticKind;
+    use crate::feedback::SourceSpan;
+    use crate::feedback::Type;
+    use crate::feedback::Violation;
     use crate::validation::test_utils::get_test_store;
 
+    #[test]
+    fn validate_yaml_err() {
```
Comment (gmuloc) [inline rust/validation/src/validation/store.rs:182]: we could add a test for: empty YAML input, comments-only YAML input and verify the result are what we want.

File: rust/validation/src/validation/store.rs [hunk 9]
```diff
@@ -120,9 +171,124 @@ pub enum StoreValidateError {
...
+                diagnostic,
+                InputDiagnostic::ParseDiagnostic(parse_diagnostic)
+                    if parse_diagnostic.kind == ParseDiagnosticKind::YamlSyntax
+                        && parse_diagnostic
+                            .to_source_span(input)
+                            .end
+                            >= parse_diagnostic.to_source_span(input).start
+            )
+        }));
+        assert!(!output.documents.is_empty());
```
Comment (gmuloc) [inline rust/validation/src/validation/store.rs:238]: we don't have a test where we end up with no document. According to my friend this could be an option:

```
let input = "*undefined_alias";
```

File: rust/validation/src/validation/store.rs [hunk 10]
```diff
@@ -120,9 +171,124 @@ pub enum StoreValidateError {
...
+                    expected: Type::Str,
+                    found: Type::Dict
+                }
+                .into()
+            },]
+        );
+    }
+
+    #[test]
+    fn validate_yaml_ok_with_coerced_data() {
```
Comment (gmuloc) [inline rust/validation/src/validation/store.rs:287]: could be interesting to make sure we can coeerce multiple documents (we can today but would help catch regressions)

File: rust/validation/src/validatable/yaml_parser_impl.rs (hunk 1 above)
Comment (ClausHolbechArista) [reply rust/validation/src/validatable/yaml_parser_impl.rs:62]: what do we want? I chose i64 as the schema type for min/max etc, but I guess we can relax it a bit for type checks and just accept any int type.

File: rust/validation/src/validation/store.rs (hunk 8 above)
Comment (ClausHolbechArista) [reply rust/validation/src/validation/store.rs:182]: are those really validation tests or yaml parser tests?

File: rust/validation/src/validation/store.rs (hunk 7 above)
Comment (ClausHolbechArista) [reply rust/validation/src/validation/store.rs:38]: well in json there will only be one result and it would break the pyvalidation signature unless we hardcode a documents.first() in there, which also seems bad in a different way.

File: rust/validation/src/validatable/yaml_parser_impl.rs (hunk 4 above)
Comment (ClausHolbechArista) [reply rust/validation/src/validatable/yaml_parser_impl.rs]: rebuilding this to preserve proper spans

File: rust/validation/src/validatable/yaml_parser_impl.rs (hunk 5 above)
Comment (ClausHolbechArista) [reply rust/validation/src/validatable/yaml_parser_impl.rs]: we are not using the spans on the coerced data so this was not caught. Rebuilding this to preserve input spans. Closing some of the overlapping comments.

File: rust/validation/src/validatable/yaml_parser_impl.rs (hunk 3 above)
Comment (ClausHolbechArista) [reply rust/validation/src/validatable/yaml_parser_impl.rs:160]: hm yeah you are right. The clone is rather expensive compared to just copying the spans, so I can make this consistent with the other to not carry over everything or change the other one to carry over?

File: rust/validation/src/validation/store.rs (hunk 7 above)
Comment (gmuloc) [reply rust/validation/src/validation/store.rs:38]: discussed and closing for now as this spearation is stronger.

File: rust/validation/src/validation/store.rs (hunk 8 above)
Comment (gmuloc) [reply rust/validation/src/validation/store.rs:182]: we would need to add test for restrict_null_value in the validation crate but not in this PR.

File: rust/validation/src/validatable/mod.rs [hunk 11]
```diff
@@ -249,13 +255,38 @@ pub trait ValidatableMappingPair<'a> {
...
+    ///
+    /// This must return a stable, human-readable key string even when
+    /// [`schema_key`](Self::schema_key) returns `None`, so validation can still
+    /// report a useful path for unexpected or ignored keys.
+    fn display_key(&self) -> Cow<'a, str>;
 
     /// Get the mapped value.
     fn value(&self) -> &'a Self::Value;
 
+    /// Build a coerced mapping entry from this pair's original key and a
```
Comment (gmuloc) [inline rust/validation/src/validatable/mod.rs:291]:
```suggestion
    /// Build a coerced mapping entry from this pair original key and a
```

File: rust/validation/src/validatable/yaml_parser_impl.rs [hunk 12]
```diff
@@ -331,32 +296,62 @@ impl<'a, 'input: 'a> Iterator for NodeMappingIter<'a, 'input> {
...
+    fn schema_key(&self) -> Option<Cow<'a, str>> {
+        schema_key_to_string(&self.pair.key)
     }
 
+    /// Return a path-friendly representation for diagnostics.
+    ///
+    /// This is deliberately broader than `schema_key` so numeric, boolean,
+    /// null, and complex YAML keys can still produce a useful validation path.
+    fn display_key(&self) -> Cow<'a, str> {
+        display_key_to_string(&self.pair.key)
```
Comment (gmuloc) [inline rust/validation/src/validatable/yaml_parser_impl.rs]: wondering if the separate helper is necessary given that today it is only used here it seems. (though I may be missing future usage)

File: rust/validation/src/validatable/yaml_parser_impl.rs [hunk 13]
```diff
@@ -331,32 +296,62 @@ impl<'a, 'input: 'a> Iterator for NodeMappingIter<'a, 'input> {
...
     fn value(&self) -> &'a Self::Value {
         &self.pair.value
     }
 
+    /// Build the YAML mapping pair for coerced output.
+    ///
+    /// String keys are rebuilt cheaply with the original key span; non-string
+    /// keys are cloned to keep their YAML value type, and the original pair
+    /// span is reused.
+    fn coerced_item(
```
Comment (gmuloc) [inline rust/validation/src/validatable/yaml_parser_impl.rs:330]: I understand that this may be for performance and because in 99% cases we expect string keys but this makes a bit of inconsistency between string keys and non string keys. Maybe at least worth stating that other metadata is technically not needed today for string keys and thats why we chose the cheap option

File: rust/validation/src/validation/dict.rs [hunk 14]
```diff
@@ -88,16 +88,15 @@ fn validate_keys<'a, M: ValidatableMapping<'a>>(
     schema: &Dict,
     input: &M,
     ctx: &mut Context,
-) -> Option<Vec<(String, <M::Value as ValidatableValue>::Coerced)>> {
+) -> Option<Vec<<M::Value as ValidatableValue>::CoercedMappingItem>> {
     let mut coerced_items = ctx.configuration.return_coerced_data.then(Vec::new);
 
     let Some(keys) = &schema.keys else {
         // No schema keys - preserve all input as-is when coercing
         if let Some(ref mut items) = coerced_items {
```
Comment (gmuloc) [inline rust/validation/src/validation/dict.rs:96]: this is never tested so we could add the following test:


```rust
  #[test]
  fn validate_no_schema_keys_preserves_input_when_coercing() {
      let schema = Dict::default();
      let input = serde_json::json!({
          "foo": 123,
          "nested": {
              "bar": true
          }
      });

      let store = get_test_store();
      let configuration = Configuration {
          return_coerced_data: true,
          ..Default::default()
      };
      let mut ctx = Context::new(&store, Some(&configuration));

      let coerced = schema.validate(&input, &mut ctx);

      assert!(ctx.result.errors.is_empty());
      assert_eq!(coerced, Some(input));
  }
```

File: rust/validation/src/validation/dict.rs
Comment (gmuloc) [inline rust/validation/src/validation/dict.rs]: we can add to this a configuration with return_coerced_data: true to make sure that removed keys are not coerced (need to assert then on the coerced value output from schema.validate

File: rust/validation/src/validatable/yaml_parser_impl.rs (hunk 6 above)
Comment (ClausHolbechArista) [reply rust/validation/src/validatable/yaml_parser_impl.rs]: keys are no longer coerced.

File: rust/validation/src/validatable/yaml_parser_impl.rs (hunk 1 above)
Comment (ClausHolbechArista) [reply rust/validation/src/validatable/yaml_parser_impl.rs:62]: emitting a new OutOfRange violation now.

File: rust/validation/src/validation/store.rs (hunk 8 above)
Comment (ClausHolbechArista) [reply rust/validation/src/validation/store.rs:182]: That would just test the yaml parser. All these would be null as discussed.

File: rust/validation/src/validatable/mod.rs (hunk 11 above)
Comment (ClausHolbechArista) [reply rust/validation/src/validatable/mod.rs:291]: ?

File: rust/validation/src/validatable/yaml_parser_impl.rs (hunk 3 above)
Comment (ClausHolbechArista) [reply rust/validation/src/validatable/yaml_parser_impl.rs:160]: let's keep it for now. It is performant enough and we should rather work on bigger things like not having to return all data as coerced etc.

File: rust/validation/src/validation/store.rs (hunk 9 above)
Comment (ClausHolbechArista) [reply rust/validation/src/validation/store.rs:238]: added

File: rust/validation/src/validation/store.rs (hunk 10 above)
Comment (ClausHolbechArista) [reply rust/validation/src/validation/store.rs:287]: enhanced this test to cover two documents.

File: rust/validation/src/validatable/mod.rs (hunk 11 above)
Comment (gmuloc) [reply rust/validation/src/validatable/mod.rs:291]: me and my sad English teacher apparently

File: rust/validation/src/validation/dict.rs [hunk 15]
```diff
@@ -117,66 +116,80 @@ fn validate_keys<'a, M: ValidatableMapping<'a>>(
...
-                        .unwrap_or_else(|| input_value.clone_to_coerced());
-                    items.push((input_key_str.to_owned(), coerced));
-                } else {
-                    let _ = key_schema.validate(input_value, ctx);
+                false // Already handled
+            } else if let Some(dynamic_keys_infos) = &dynamic_keys_infos
+                && let Some(dynamic_key_info) = dynamic_keys_infos.get(input_schema_key)
+            {
+                let key_schema = dynamic_key_info.schema;
+                if !check_deprecation(input_schema_key, key_schema, key_span, input, ctx) {
```
Comment (gmuloc) [inline rust/validation/src/validation/dict.rs:150]: we don't have tests for deprecated dynamic keys today and we will soon need to do so as we remove them so worth adding it for now:


```rust
  #[test]
  fn validate_dynamic_key_deprecated_ok() {
      let schema: Dict = Dict::deserialize(serde_json::json!({
          "keys": {
              "my_dynamic_keys": {
                  "type": "list",
                  "items": {
                      "type": "dict",
                      "keys": {
                          "key": {
                              "type": "str"
                          }
                      }
                  }
              }
          },
          "dynamic_keys": {
              "my_dynamic_keys.key": {
                  "type": "int",
                  "deprecation": {
                      "warning": true,
                      "remove_in_version": "1.2.3"
                  }
              }
          },
          "allow_other_keys": true
      }))
      .unwrap();

      let input = serde_json::json!({
          "my_dynamic_keys": [{"key": "dynkey1"}],
          "dynkey1": 5
      });

      let store = get_test_store();
      let mut ctx = Context::new(&store, None);
      let _ = schema.validate(&input, &mut ctx);

      assert!(ctx.result.errors.is_empty());
      assert_eq!(
          ctx.result.warnings,
          vec![Feedback {
              path: vec!["dynkey1".into()].into(),
              span: None,
              issue: WarningIssue::Deprecated(Deprecated {
                  path: vec!["dynkey1".into()].into(),
                  replacement: None.into(),
                  version: Some("1.2.3".into()).into(),
                  url: None.into(),
              })
          }]
      );
  }
```

Comment (ClausHolbechArista) [reply rust/validation/src/validation/dict.rs:150]: no, this will be changed in a later pr where all keys (dynamic, prefix or static) will be handles with the same logic.

Comment (ClausHolbechArista) [reply rust/validation/src/validation/dict.rs:150]: this is general validation coverage.

File: rust/validation/src/validation/dict.rs
Comment (ClausHolbechArista) [reply rust/validation/src/validation/dict.rs]: why do we care? removed keys will block further execution.

File: rust/validation/src/validatable/yaml_parser_impl.rs (hunk 13 above)
Comment (ClausHolbechArista) [reply rust/validation/src/validatable/yaml_parser_impl.rs:330]: it is not needed for any key. You said we didn't care in your previous comments but you seem to care :) I will clone all of it now so we can move on.

**Review by gmuloc — APPROVED** [review]
LGTM other than the nits about coverage that can be tackled in another PR

File: rust/validation/src/validatable/yaml_parser_impl.rs (hunk 12 above)
Comment (ClausHolbechArista) [reply rust/validation/src/validatable/yaml_parser_impl.rs]: removed helper

File: rust/validation/src/validation/dict.rs (hunk 14 above)
Comment (ClausHolbechArista) [reply rust/validation/src/validation/dict.rs:96]: added

File: rust/validation/src/validation/dict.rs (hunk 15 above)
Comment (ClausHolbechArista) [reply rust/validation/src/validation/dict.rs:150]: added test

File: rust/validation/src/validation/dict.rs
Comment (ClausHolbechArista) [reply rust/validation/src/validation/dict.rs]: added test
